tools/skbstat.py

```diff
@@ -0,0 +1,1079 @@
+        addr_filter += "\tfilter->addr6[%d][3] = 0x%x;\n" % (pos, a3)
+
+        addr_filter += "\tfilter->mask6[%d][0] = 0x%x;\n" % (pos, m0)
+        addr_filter += "\tfilter->mask6[%d][1] = 0x%x;\n" % (pos, m1)
+        addr_filter += "\tfilter->mask6[%d][2] = 0x%x;\n" % (pos, m2)
+        addr_filter += "\tfilter->mask6[%d][3] = 0x%x;\n" % (pos, m3)
+
+    return addr_filter, net.version
+
+flags_filter = ""
+
+saddr_filter = ""
+if args.saddr:
+    saddr_filter, ip_version = init_addr_filter(args.saddr, 0)
+
+daddr_filter = ""
+if args.daddr:
+    daddr_filter, new_version = init_addr_filter(args.daddr, 1)
+    ip_version = init_ip_version(ip_version, new_version)
+
+if args.addr:
+    flags_filter = "filter->flags |= SKB_ADDR_MATCH_1;\n"
+    daddr_filter = ""
+    saddr_filter, ip_version = init_addr_filter(args.addr, 0)
+
+ip_version_filter = "filter->ip_version = %d;\n" % ip_version
+
+def check_port(port):
+    if port < 0 or port > 65535:
+        print("invalid port %d" % port)
+        exit()
+
+def check_port_range(port):
+    if int(port[0]) > int(port[1]):
+        print("invalid port range %d-%d" % (int(port[0]), int(port[1])))
+        exit()
+
+def init_port_filter(port_str, name):
+    ports = port_str.split('-')
+    port = int(ports[0])
+    check_port(port)
+
+    port_filter = "filter->%s[0] = htons(%u);\n" % (name, port)
+
+    if len(ports) > 1:
+        port = int(ports[1])
+        check_port(port)
+        check_port_range(ports)
+
+    port_filter += "\tfilter->%s[1] = htons(%u);\n" % (name, port)
+
+    return port_filter
+
+sport_filter = ""
+if args.sport:
+    sport_filter = init_port_filter(args.sport, "sport_range")
+
+dport_filter = ""
+if args.dport:
+    dport_filter = init_port_filter(args.dport, "dport_range")
+
+if args.port:
+    if flags_filter:
+        flags_filter += "\t"
+    flags_filter += "filter->flags |= SKB_PORT_MATCH_1;\n"
+
+    dport_filter = ""
+    sport_filter = init_port_filter(args.port, "sport_range")
+
+def proto_str2int(proto):
+    if proto.lower() == "icmp":
+        return 1
+    if proto.lower() == "tcp":
+        return 6
+    if proto.lower() == "udp":
+        return 17
+    if proto.lower() == "icmp6":
+        return 58
+    if proto.isdigit():
+        proto_num = int(proto)
+        if proto_num >= 0 and proto_num <= 255:
+            return proto_num
+    return 0
+
+proto = 0
+proto_filter = ""
+if args.proto:
+    proto = proto_str2int(args.proto)
+    proto_filter += "filter->proto = %d;\n" % proto
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <uapi/linux/ip.h>
+#include <uapi/linux/ipv6.h>
+#include <uapi/linux/if_ether.h>
+#include <net/sock.h>
+#include <bcc/proto.h>
+
+BPF_TABLE("percpu_hash", u32, u64, drop_stacks, 256);
+BPF_TABLE("percpu_hash", u32, u64, consume_stacks, 256);
+BPF_TABLE("percpu_hash", u64, u64, input, 256);
+BPF_TABLE("percpu_hash", u64, u64, forward, 256);
```

Comment:
You forgot to use BPF_PERCPU_HASH, since another revision is needed, suggest you use BPF_PERCPU_HASH since it is simpler.

Reply:
I have concern with older bcc version which has no BPF_PERCPU_HASH. This tool will not work when copied to some machine which only has old bcc version.

Reply:
okay, I won't insist on this one.